### PR TITLE
`ThemeSwitcher`, `LocaleSwitcher` and generic `Provider` widgets

### LINKED ISSUE #256: Provide a widget that injects a function to enable consumers to change theme
**Enhancement**

Currently it takes a significant amount of boilerplate to inject a function that can be used to change the current theme into the widget tree:

```ts
// main.ts
import renderer from "@dojo/framework/widget-core/vdom";
import { w } from "@dojo/framework/widget-core/d";
import App from "./widgets/App";
import { Registry } from "@dojo/framework/widget-core/Registry";
import { registerThemeInjector, Theme } from "@dojo/framework/widget-core/mixins/Themed";
import myTheme from './my-theme';

const registry = new Registry();
const injector = registerThemeInjector(myTheme, registry);

registry.defineInjector("update-theme", invalidator => {
  function themeSwitcher(theme: Theme) {
    injector.set({ theme });
  }
  return () => themeSwitcher;
});

const r = renderer(() => w(App, {}));
r.mount({ registry });

// ThemeSwitcherContainer.ts
// Used in the application somewhere to inject the theme switcher function

import Container from "@dojo/framework/widget-core/Container";
import ThemeSwitcher from "./ThemeSwitcher";

export default Container(ThemeSwitcher, "update-theme", {
  getProperties(injector) {
    return {
      themeSwitcher: injector
    };
  }
});
```

It would be helpful if Dojo could provide a customised widget that injects a function that can be to switch themes into an application widget tree.

The widget should be virtual and have no visual output, but provide an API to for the user to render widgets and nodes with access to the update theme function. This would remove the need for a user to register a new injector or create a custom container to get access to the update theme function.

### LINKED ISSUE #257: Provide a widget that injects a function to enable consumers to change locale
**Enhancement**

Currently it takes a significant amount of boilerplate to inject a function that can be used to change the locale into the widget tree.

It would be helpful if Dojo could provide a customised widget that injects a function that can be to switch locale into an application widget tree.

The widget should be virtual and have no visual output, but provide an API to for the user to render widgets and nodes with access to the update locale function. This would remove the need for a user to register a new injector or create a custom container to get access to the update locale function.

Related to #256

### LINKED ISSUE #258: Generic Provider widget for injecting registry items into the widget tree
**Enhancement**

A generic widget that uses a render property to inject injectors into the widget tree, an alternative to using the `Container` HOC pattern.

Example:

```ts
w(Provider, { registryLabel: 'my-injector', renderer: (injector) => {
  return v('div', [ injector.value ]);
}})
```

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

This pull requests adds a collection of similar widgets that can be used to work with injector values from an applications registry.

* ThemeSwitcher - Injects a function that can be used to change an the theme set in the theme injector
* LocaleSwitcher - Injects a function that can be used to change an the locale data set in the i18n injector
* Provider - A generic provider that enables consumers to work with injectors.

Resolves #258 
Resolves #257 
Resolves #256

### CODE REVIEW

File: src/widget-core/README.md [hunk 1]
```diff
@@ -557,6 +557,85 @@ render() {
 }
 ```
 
+#### Managing Themes throughout an Application
```
Comment (dylans) [inline src/widget-core/README.md]:
```suggestion
#### Managing Themes Throughout an Application
```

File: src/widget-core/README.md [hunk 2]
```diff
@@ -557,6 +557,85 @@ render() {
...
+
+To manage themes throughout an application, the `ThemedMixin` leverages the [`Containers` and `Injectors`](#containers--injectors) to set an application's locale and inject the locale data into all widgets using the `ThemedMixin`. When the locale is updated in the theme `Injector` all themed widgets will be invalidated and re-rendered with the updated locale.
+
+`@dojo/framework/widget-core/mixins/Themed` exposes a convenience method, `registerThemeInjector` for registering the `theme` injector with a registry.
```
Comment (dylans) [inline src/widget-core/README.md]:
```suggestion
`@dojo/framework/widget-core/mixins/Themed` exposes a convenience method, `registerThemeInjector`, for registering the `theme` injector with a registry.
```

File: src/widget-core/README.md [hunk 3]
```diff
@@ -557,6 +557,85 @@ render() {
...
+import App from './App';
+
+const registry = new Registry();
+registerThemeInjector(myTheme, registry);
+
+const r = renderer(() => w(App, {}));
+r.mount({ registry });
+```
+
+#### Changing Theme
```
Comment (dylans) [inline src/widget-core/README.md]:
```suggestion
#### Changing the Theme
```

File: src/widget-core/README.md [hunk 4]
```diff
@@ -557,6 +557,85 @@ render() {
...
+
+To change a theme a widget `ThemeSwitcher` is available from `@dojo/framework/widget-core/mixins/Themed`, the `ThemeSwitcher` widget has a `renderer` property that injects an `updateTheme` function and returns `DNode | DNode[]` to render.
```
Comment (dylans) [inline src/widget-core/README.md]:
```suggestion
To change the theme a widget `ThemeSwitcher` is available from `@dojo/framework/widget-core/mixins/Themed`. The `ThemeSwitcher` widget has a `renderer` property that injects an `updateTheme` function and returns `DNode | DNode[]` to render.
```

File: src/widget-core/README.md [hunk 5]
```diff
@@ -557,6 +557,85 @@ render() {
...
+
+##### Properties
+
+-   `renderer`: (updateTheme(theme: Theme) => void): DNode | DNode[]
+-   The renderer that is called with the `updateTheme` function and returns `DNode | DNode[]` that will be rendered
```
Comment (dylans) [inline src/widget-core/README.md]: Should this be further indented? Should renderer be in backticks?

File: src/widget-core/README.md [hunk 6]
```diff
@@ -664,6 +743,83 @@ export class MyWidget extends WidgetBase {
...
+```
+
+#### Changing LocaleData
+
+When using the `Registry` to inject i18n properties to widgets using the `I18nMixin`, `@dojo/framework/widget-core/mixins/I18n` provides a widget, `LocaleSwitcher` that can be used to inject a function for changing the applications locale. The widget has a `renderer` property that injects an `updateLocale` function and returns `DNode | DNode[]` to render.
+
+##### Properties
+
+-   `renderer`: (updateLocale(localeData: LocaleData) => void): DNode | DNode[]
+-   The renderer that is called with the `updateLocale` function and returns `DNode | DNode[]` that will be rendered
```
Comment (dylans) [inline src/widget-core/README.md]: Same notes as the theme section

File: src/widget-core/README.md [hunk 7]
```diff
@@ -1068,7 +1224,22 @@ registry.defineInjector('my-injector', (invalidator) => {
 });
 ```
 
-To connect the registered `payload` to a widget, we can use the `Container` HOC (higher order component) provided by `widget-core`. The `Container` accepts a widget `constructor`, `injector` label, and `getProperties` mapping function as arguments and returns a new class that returns the passed widget from its `render` function.
+To connect (inject) the registered values to a widget there are two options. The first is the `Provider` which is a widget that is designed to be used inline the a widgets `render`. The `Provider` takes a `registryLabel` property and a `renderer` property that received the values from the registry. The second is the `Container` HOC (higher order component).
```
Comment (dylans) [inline src/widget-core/README.md]:
```suggestion
To connect (inject) the registered values to a widget there are two options. The first is the `Provider` which is a widget that is designed to be used inline the widgets `render`. The `Provider` takes a `registryLabel` property and a `renderer` property that received the values from the registry. The second is the `Container` HOC (higher order component).
```

File: src/widget-core/README.md [hunk 8]
```diff
@@ -557,6 +557,85 @@ render() {
 }
 ```
 
+#### Managing Themes Throughout an Application
+
+To manage themes throughout an application, the `ThemedMixin` leverages the [Injectors, Providers & Containers](#injectors-providers--containers) to set an application's locale and inject the locale data into all widgets using the `ThemedMixin`. When the locale is updated in the theme `Injector` all themed widgets will be invalidated and re-rendered with the updated locale.
```
Comment (matt-gadd) [inline src/widget-core/README.md]: This mentions locales rather than themes near the end

Comment (agubler) [reply src/widget-core/README.md]: copy paste special, thanks @matt-gadd

File: src/widget-core/README.md [hunk 9]
```diff
@@ -664,6 +743,83 @@ export class MyWidget extends WidgetBase {
...
+-   `renderer`: (updateLocale(localeData: LocaleData) => void): DNode | DNode[]
+    -   The `renderer` that is called with the `updateLocale` function and returns `DNode | DNode[]` that will be rendered
+-   `registryLabel`(optional): string
+    -   The registry label used to register the i18n injector. When using the `registerI18nInjector` this does not need to be set.
+
+##### Example Usage
+
+```ts
+import WidgetBase from '@dojo/framework/widget-core/WidgetBase';
+import I18ndMixin, { LocaleSwitcher, UpdateLocale } from '@dojo/framework/widget-core/mixins/I18n';
```
Comment (sbinge) [inline src/widget-core/README.md]: Minor typo, here & in the `MyApp` definition - `s/I18ndMixin/I18nMixin/`

File: src/widget-core/README.md (hunk 8 above)
Comment (sbinge) [reply src/widget-core/README.md]: FYI, think this is still an issue in the PR's current readme.

Similar to i18n, I'm busy porting this over (with s/locale/theme/) to the theme ref guide so this section will shortly be redundant anyway.

**Comment by agubler** [comment]
Closing as this functionality is going to be built out with the upcoming render middleware.

**Pull request closed without merging**